xwiki-commons-core/xwiki-commons-component/xwiki-commons-component-default/src/main/java/org/xwiki/component/embed/EmbeddableComponentManager.java

```diff
@@ -196,27 +412,45 @@ public <T> T getInstance(Type roleType, String roleHint) throws ComponentLookupE
 
         ComponentEntry<T> componentEntry = (ComponentEntry<T>) getComponentEntry(roleType, roleHint);
 
-        if (componentEntry != null) {
-            try {
-                instance = getComponentInstance(componentEntry);
-            } catch (Throwable e) {
-                throw new ComponentLookupException(
-                    String.format("Failed to lookup component [%s] identified by type [%s] and hint [%s]",
-                        componentEntry.descriptor.getImplementation().getName(), roleType, roleHint),
-                    e);
-            }
-        } else {
+        if (componentEntry == null) {
             if (getParent() != null) {
                 instance = getParent().getInstance(roleType, roleHint);
             } else {
                 throw new ComponentLookupException(
                     "Can't find descriptor for the component with type [" + roleType + "] and hint [" + roleHint + "]");
             }
+        } else if (getParent() != null) {
+            // Get parent descriptor
+            ComponentDescriptor<T> parentDescriptor = getParent().getComponentDescriptor(roleType, roleHint);
+
+            if (parentDescriptor != null
+                && parentDescriptor.getRoleHintPriority() < componentEntry.descriptor.getRoleHintPriority()) {
+                // The parent component has priority over the current one
+                instance = getParent().getInstance(roleType, roleHint);
+            } else {
+                // The current component has priority over the parent one
+                instance = getInstance(componentEntry);
+            }
+        } else {
+            instance = getInstance(componentEntry);
         }
 
         return instance;
     }
 
+    private <T> T getInstance(ComponentEntry<T> componentEntry) throws ComponentLookupException
+    {
+        try {
+            return getComponentInstance(componentEntry);
+        } catch (Throwable e) {
```

Comment:
Isn't sonar complaining for that? https://rules.sonarsource.com/java/RSPEC-1181

Reply:
Yeah, I moved code all over the place. I guess there is a good chance sonar too thinks it's new code.